File: scripts/position_test.py
#!/usr/bin/env python3
"""
Linux-side position controller using TORQUE mode with fast CAN feedback.

Uses FAST_FRAME (TPDO4 at 0x481) for position feedback,
sends torque commands via PDO3 (RPDO3 at 0x401).

Usage: python3 position_test.py [target_degrees] [duration_seconds]
"""

import socket
import struct
import time
import math
import sys

DEVICE_ID = 1
CAN_INTERFACE = "can0"

# Motor parameters
TORQUE_CONSTANT = 0.08958

# Control parameters
TORQUE_LIMIT = 0.3       # Nm
I_LIMIT = 5.0
FAST_FRAME_HZ = 500

# PD gains
KP = 1.5       # Nm/rad
KD = 0.02      # Nm/(rad/s)
KI = 0.3       # Nm/(rad*s)

# Trajectory limits
MAX_VEL = 5.0         # rad/s max trajectory velocity
MAX_ACCEL = 20.0      # rad/s^2 max trajectory acceleration

# CAN IDs
FUNC_NMT = 0
FUNC_RECEIVE_PDO_3 = 8  # 0x401 = [position_target, torque_target]
FUNC_TRANSMIT_PDO_4 = 9  # 0x481 = fast frame [position, velocity]
FUNC_TRANSMIT_SDO = 0xB
FUNC_RECEIVE_SDO = 0xC
FUNC_FLASH = 0xD
FUNC_HEARTBEAT = 0xE


def make_can_id(func, dev_id):
    return (func << 7) | dev_id


class CANBus:
    def __init__(self, interface):
        self.sock = socket.socket(socket.PF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
        self.sock.bind((interface,))

    def send(self, can_id, data=b""):
        dlc = len(data)
        self.sock.send(struct.pack("<IB3x8s", can_id, dlc, data.ljust(8, b'\x00')))

    def recv(self, timeout_s=0.1):
        """Receive a CAN frame with timeout. Returns (can_id, data) or None."""
        self.sock.settimeout(timeout_s)
        try:
            raw = self.sock.recv(16)
            can_id, dlc = struct.unpack("<IB3x", raw[:8])
            return (can_id & 0x7FF, raw[8:8 + dlc])
        except socket.timeout:
            return None

    def drain(self):
        """Drain all pending frames."""
        self.sock.settimeout(0)
        try:
            while True:
                self.sock.recv(16)
        except (BlockingIOError, socket.timeout):
            pass

    def sdo_read_float(self, param_id):
        self.send(make_can_id(FUNC_RECEIVE_SDO, DEVICE_ID),
                  struct.pack('<BH', 0x40, param_id))
        expected = make_can_id(FUNC_TRANSMIT_SDO, DEVICE_ID)
        deadline = time.time() + 0.5
        while time.time() < deadline:
            r = self.recv(0.05)
            if r and r[0] == expected and len(r[1]) >= 4:
                return struct.unpack('<f', r[1][:4])[0]
        return None

    def sdo_write_float(self, param_id, value):
        iv = struct.unpack('<I', struct.pack('<f', value))[0]
        self.send(make_can_id(FUNC_RECEIVE_SDO, DEVICE_ID),
                  struct.pack('<BHxI', 0x20, param_id, iv))

    def sdo_write_u32(self, param_id, value):
        self.send(make_can_id(FUNC_RECEIVE_SDO, DEVICE_ID),
                  struct.pack('<BHxI', 0x20, param_id, value))

    def sdo_read_u32(self, param_id):
        self.send(make_can_id(FUNC_RECEIVE_SDO, DEVICE_ID),
                  struct.pack('<BH', 0x40, param_id))
        expected = make_can_id(FUNC_TRANSMIT_SDO, DEVICE_ID)
        deadline = time.time() + 0.5
        while time.time() < deadline:
            r = self.recv(0.05)
            if r and r[0] == expected and len(r[1]) >= 4:
                return struct.unpack('<I', r[1][:4])[0]
        return None

    def send_nmt(self, mode):
        self.send(make_can_id(FUNC_NMT, 0), struct.pack('BB', mode, DEVICE_ID))

    def send_heartbeat(self):
        self.send(make_can_id(FUNC_HEARTBEAT, DEVICE_ID), b'')

    def send_torque(self, torque):
        """Send torque command via PDO3: [position_target(ignored), torque_target]."""
        self.send(make_can_id(FUNC_RECEIVE_PDO_3, DEVICE_ID),
                  struct.pack('<ff', 0.0, torque))

    def close(self):
        self.sock.close()


def main():
    target_deg = float(sys.argv[1]) if len(sys.argv) > 1 else 90.0
    duration = float(sys.argv[2]) if len(sys.argv) > 2 else 5.0

    can = CANBus(CAN_INTERFACE)

    # Reset
    print("Resetting motor...")
    can.sdo_write_u32(0x014, 0)
    can.send_heartbeat()
    can.send_nmt(0x00)  # DISABLED
    time.sleep(0.5)
    for _ in range(5):
        can.send_heartbeat()
        time.sleep(0.1)
    can.send_nmt(0x01)  # IDLE
    time.sleep(0.5)
    for _ in range(5):
        can.send_heartbeat()
        time.sleep(0.1)

    mode = can.sdo_read_u32(0x010)
    err = can.sdo_read_u32(0x014)
    print(f"  mode=0x{mode:02X}, error=0x{err:08X}")

    # Configure
    can.sdo_write_float(0x074, I_LIMIT)
    can.sdo_write_float(0x030, TORQUE_LIMIT)
    can.sdo_write_u32(0x00C, FAST_FRAME_HZ)
    time.sleep(0.2)

    # Read starting position
    can.drain()
    pos_start = can.sdo_read_float(0x134)
    if pos_start is None:
        print("ERROR: Cannot read position")
        can.close()
        return 1

    target_rad = pos_start + math.radians(target_deg)
    print(f"Start: {math.degrees(pos_start):.1f} deg")
    print(f"Target: {math.degrees(target_rad):.1f} deg  (+{target_deg} deg)")
    print(f"Gains: Kp={KP}, Kd={KD}, Ki={KI}")
    print()

    # Enter TORQUE mode
    can.send_torque(0.0)
    time.sleep(0.05)
    can.send_nmt(0x11)  # TORQUE
    time.sleep(0.3)
    can.send_heartbeat()
    can.drain()

    mode = can.sdo_read_u32(0x010)
    err = can.sdo_read_u32(0x014)
    print(f"  Torque mode: mode=0x{mode:02X}, error=0x{err:08X}")
    if mode != 0x11:
        print("ERROR: Failed to enter TORQUE mode!")
        can.close()
        return 1

    # Control state
    prev_pos = pos_start
    prev_time = time.time()
    vel_filtered = 0.0
    vel_alpha = 0.15
    integral = 0.0
    integral_limit = TORQUE_LIMIT * 0.5

    # Trajectory state
    traj_pos = pos_start   # current trajectory setpoint
    traj_vel = 0.0         # current trajectory velocity

    fast_frame_id = make_can_id(FUNC_TRANSMIT_PDO_4, DEVICE_ID)
    start = time.time()
    last_print = 0
    last_hb = 0
    frame_count = 0

    print(f"{'time':>5s}  {'error':>8s}  {'pos':>9s}  {'setpt':>9s}  {'vel':>8s}  {'torque':>8s}")
    print("-" * 58)

    try:
        while time.time() - start < duration:
            frame = can.recv(timeout_s=0.01)
            now = time.time()

            # Heartbeat every 50ms
            if now - last_hb >= 0.05:
                can.send_heartbeat()
                last_hb = now

            if frame is None:
                continue

            can_id, data = frame
            if can_id != fast_frame_id or len(data) < 8:
                continue

            frame_count += 1
            pos = struct.unpack('<f', data[0:4])[0]

            # Velocity estimation
            dt = now - prev_time
            if dt < 0.0005:
                continue
            vel_raw = (pos - prev_pos) / dt
            vel_filtered = vel_alpha * vel_raw + (1 - vel_alpha) * vel_filtered
            prev_pos = pos
            prev_time = now

            # Trajectory generator: ramp traj_pos toward target_rad
            pos_to_go = target_rad - traj_pos
            direction = 1.0 if pos_to_go > 0 else -1.0

            # Deceleration distance: v^2 / (2*a)
            decel_dist = traj_vel * traj_vel / (2.0 * MAX_ACCEL)

            if abs(pos_to_go) <= 0.001 and abs(traj_vel) < 0.1:
                # Close enough, hold position
                traj_pos = target_rad
                traj_vel = 0.0
            elif abs(pos_to_go) <= decel_dist:
                # Decelerate
                traj_vel -= direction * MAX_ACCEL * dt
                if direction * traj_vel < 0:
                    traj_vel = 0.0
            else:
                # Accelerate toward target (up to max vel)
                traj_vel += direction * MAX_ACCEL * dt
                traj_vel = max(-MAX_VEL, min(MAX_VEL, traj_vel))

            traj_pos += traj_vel * dt

            # PD+I controller tracks trajectory setpoint
            pos_error = traj_pos - pos
            vel_error = traj_vel - vel_filtered

            integral += KI * pos_error * dt
            integral = max(-integral_limit, min(integral_limit, integral))

            torque = KP * pos_error + KD * vel_error + integral
            torque = max(-TORQUE_LIMIT, min(TORQUE_LIMIT, torque))

            # Send torque via PDO3
            can.send_torque(torque)

            # Print
            elapsed = now - start
            if elapsed - last_print >= 0.25:
                print(f"{elapsed:5.1f}s  {math.degrees(pos_error):+7.1f}deg  "
                      f"{math.degrees(pos):8.1f}deg  "
                      f"{math.degrees(traj_pos):8.1f}deg  "
                      f"{vel_filtered:+7.1f}r/s  "
                      f"{torque:+.4f}Nm")
                last_print = elapsed

    except KeyboardInterrupt:
        print("\nInterrupted!")

    # Stop
    print("\nStopping...")
    can.send_torque(0.0)
    time.sleep(0.1)
    can.send_nmt(0x02)  # DAMPING

    elapsed = time.time() - start
    print(f"Frames: {frame_count} ({frame_count/max(elapsed,0.1):.0f} Hz)")

    final_pos = can.sdo_read_float(0x134)
    if final_pos is not None:
        print(f"Final: {math.degrees(final_pos):.1f} deg  (error: {math.degrees(target_rad - final_pos):+.1f} deg)")

    print("Done.")
    can.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/position_test.cpp
/**
 * Linux-side position controller using TORQUE mode with fast CAN feedback.
 *
 * Motor is connected through a 14:1 gearbox.
 * Encoder is on the motor shaft (not gearbox output).
 * Target angle is specified in OUTPUT shaft degrees.
 * Motor must turn 14x more than the output.
 *
 * Uses FAST_FRAME (TPDO4 at 0x481) for motor position feedback at 500 Hz,
 * sends torque commands via PDO3 (RPDO3 at 0x401).
 * Trapezoidal velocity trajectory planner with PD+I tracking controller.
 *
 * Usage: ./position_test [output_degrees] [duration_seconds]
 */

#include "can_interface.hpp"

#include <iostream>
#include <chrono>
#include <thread>
#include <csignal>
#include <cmath>
#include <cstring>
#include <iomanip>
#include <optional>
#include <algorithm>

volatile sig_atomic_t g_running = 1;
void signalHandler(int) { g_running = 0; }

// CAN protocol
constexpr uint8_t DEVICE_ID = 1;
constexpr uint32_t makeCanId(uint32_t func, uint32_t dev) { return (func << 7) | dev; }

constexpr uint32_t CAN_ID_NMT        = makeCanId(0x0, 0);          // 0x000
constexpr uint32_t CAN_ID_PDO3_TX    = makeCanId(0x8, DEVICE_ID);  // 0x401 - torque cmd
constexpr uint32_t CAN_ID_FAST_FRAME = makeCanId(0x9, DEVICE_ID);  // 0x481 - fast feedback
constexpr uint32_t CAN_ID_SDO_TX     = makeCanId(0xB, DEVICE_ID);  // 0x581 - SDO response
constexpr uint32_t CAN_ID_SDO_RX     = makeCanId(0xC, DEVICE_ID);  // 0x601 - SDO request
constexpr uint32_t CAN_ID_HEARTBEAT  = makeCanId(0xE, DEVICE_ID);  // 0x701

// Gearbox
constexpr float GEAR_RATIO = 14.0f;   // motor turns per output turn

// Control parameters
constexpr float TORQUE_LIMIT   = 0.3f;    // Nm (at motor shaft)
constexpr float I_LIMIT        = 5.0f;    // A
constexpr uint32_t FAST_HZ     = 500;

// PD+I gains (all in motor-shaft units)
constexpr float KP = 0.3f;    // Nm/rad — gentle: 0.3 Nm per rad of motor error
constexpr float KD = 0.02f;   // Nm/(rad/s) — damping at motor shaft
constexpr float KI = 0.05f;   // Nm/(rad*s) — slow integral

// Trajectory limits (motor shaft)
constexpr float MAX_VEL   = 20.0f;   // rad/s motor (~100 deg/s output)
constexpr float MAX_ACCEL = 30.0f;   // rad/s^2 motor

// SDO parameter IDs
constexpr uint16_t PARAM_MODE       = 0x010;
constexpr uint16_t PARAM_ERROR      = 0x014;
constexpr uint16_t PARAM_FAST_HZ    = 0x00C;
constexpr uint16_t PARAM_TORQUE_LIM = 0x030;
constexpr uint16_t PARAM_I_LIM      = 0x074;
constexpr uint16_t PARAM_POSITION   = 0x134;

static double now_s() {
    using namespace std::chrono;
    return duration<double>(steady_clock::now().time_since_epoch()).count();
}

class MotorCAN {
    CANInterface& can_;
public:
    explicit MotorCAN(CANInterface& c) : can_(c) {}

    void sendNMT(uint8_t mode) {
        uint8_t d[2] = {mode, DEVICE_ID};
        can_.send(CAN_ID_NMT, d, 2);
    }
    void sendHeartbeat() { can_.send(CAN_ID_HEARTBEAT, nullptr, 0); }

    void sendTorque(float torque) {
        uint8_t d[8] = {};
        float zero = 0.0f;
        std::memcpy(d, &zero, 4);
        std::memcpy(d + 4, &torque, 4);
        can_.send(CAN_ID_PDO3_TX, d, 8);
    }

    bool sdoWriteFloat(uint16_t param, float val) {
        uint8_t d[8] = {};
        d[0] = 0x20;
        std::memcpy(d + 1, &param, 2);
        std::memcpy(d + 4, &val, 4);
        return can_.send(CAN_ID_SDO_RX, d, 8);
    }

    bool sdoWriteU32(uint16_t param, uint32_t val) {
        uint8_t d[8] = {};
        d[0] = 0x20;
        std::memcpy(d + 1, &param, 2);
        std::memcpy(d + 4, &val, 4);
        return can_.send(CAN_ID_SDO_RX, d, 8);
    }

    std::optional<uint32_t> sdoReadU32(uint16_t param) {
        uint8_t d[3];
        d[0] = 0x40;
        std::memcpy(d + 1, &param, 2);
        can_.send(CAN_ID_SDO_RX, d, 3);
        auto deadline = now_s() + 0.5;
        while (now_s() < deadline) {
            auto f = can_.receive(50);
            if (f && f->id == CAN_ID_SDO_TX && f->data.size() >= 4) {
                uint32_t v;
                std::memcpy(&v, f->data.data(), 4);
                return v;
            }
        }
        return std::nullopt;
    }

    std::optional<float> sdoReadFloat(uint16_t param) {
        auto v = sdoReadU32(param);
        if (!v) return std::nullopt;
        float f;
        uint32_t raw = *v;
        std::memcpy(&f, &raw, 4);
        return f;
    }

    void drain() { while (can_.receive(0)) {} }
    std::optional<CANFrame> receive(int ms) { return can_.receive(ms); }
};

int main(int argc, char* argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    float output_deg = (argc > 1) ? std::stof(argv[1]) : 90.0f;
    float duration   = (argc > 2) ? std::stof(argv[2]) : 5.0f;

    float motor_deg = output_deg * GEAR_RATIO;

    std::cout << "Position Test (C++) — Torque mode + trajectory planner\n"
              << "Gear ratio: " << GEAR_RATIO << ":1\n"
              << "Output target: " << output_deg << " deg\n"
              << "Motor target:  " << motor_deg << " deg\n"
              << "Duration: " << duration << " s\n";

    CANInterface can;
    if (!can.open("can0")) {
        std::cerr << "Failed to open CAN\n";
        return 1;
    }
    MotorCAN motor(can);

    // ---- Reset ----
    std::cout << "Resetting motor...\n";
    motor.sdoWriteU32(PARAM_ERROR, 0);
    motor.sendHeartbeat();
    motor.sendNMT(0x00);  // DISABLED
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    for (int i = 0; i < 5; i++) {
        motor.sendHeartbeat();
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    motor.sendNMT(0x01);  // IDLE
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    for (int i = 0; i < 5; i++) {
        motor.sendHeartbeat();
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    {
        auto m = motor.sdoReadU32(PARAM_MODE);
        auto e = motor.sdoReadU32(PARAM_ERROR);
        std::cout << "  mode=0x" << std::hex;
        if (m) std::cout << *m; else std::cout << "??";
        std::cout << ", error=0x";
        if (e) std::cout << *e; else std::cout << "??";
        std::cout << std::dec << "\n";
    }

    // ---- Configure ----
    motor.sdoWriteFloat(PARAM_I_LIM, I_LIMIT);
    motor.sdoWriteFloat(PARAM_TORQUE_LIM, TORQUE_LIMIT);
    motor.sdoWriteU32(PARAM_FAST_HZ, FAST_HZ);
    std::this_thread::sleep_for(std::chrono::milliseconds(200));

    // ---- Read start position (motor shaft) ----
    motor.drain();
    auto pos0 = motor.sdoReadFloat(PARAM_POSITION);
    if (!pos0) {
        std::cerr << "ERROR: Cannot read position\n";
        return 1;
    }
    float pos_start  = *pos0;
    float target_rad = pos_start + motor_deg * static_cast<float>(M_PI) / 180.0f;

    std::cout << std::fixed << std::setprecision(1)
              << "Motor start:  " << (pos_start * 180.0f / M_PI) << " deg\n"
              << "Motor target: " << (target_rad * 180.0f / M_PI) << " deg  (+"
              << motor_deg << " deg motor = " << output_deg << " deg output)\n"
              << "Gains: Kp=" << KP << " Kd=" << KD << " Ki=" << KI << "\n"
              << "Traj:  max_vel=" << MAX_VEL << " rad/s  max_accel=" << MAX_ACCEL << " rad/s^2\n\n";

    // ---- Enter TORQUE mode ----
    motor.sendTorque(0.0f);
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    motor.sendNMT(0x11);
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    motor.sendHeartbeat();
    motor.drain();

    {
        auto m = motor.sdoReadU32(PARAM_MODE);
        auto e = motor.sdoReadU32(PARAM_ERROR);
        std::cout << "  Torque mode: mode=0x" << std::hex;
        if (m) std::cout << *m; else std::cout << "??";
        std::cout << ", error=0x";
        if (e) std::cout << *e; else std::cout << "??";
        std::cout << std::dec << "\n";
        if (!m || *m != 0x11) {
            std::cerr << "ERROR: Failed to enter TORQUE mode!\n";
            return 1;
        }
    }

    // ---- Control loop state ----
    float prev_pos    = pos_start;
    double prev_time  = now_s();
    float vel_filt    = 0.0f;
    float integral    = 0.0f;
    float traj_pos    = pos_start;
    float traj_vel    = 0.0f;

    constexpr float VEL_ALPHA   = 0.15f;
    constexpr float INT_LIMIT   = TORQUE_LIMIT * 0.5f;

    double t0         = now_s();
    double last_print = 0.0;
    double last_hb    = 0.0;
    uint32_t frames   = 0;

    // Header: show output shaft degrees
    std::cout << std::right
              << std::setw(5)  << "time"   << "  "
              << std::setw(9)  << "out_err" << "  "
              << std::setw(9)  << "out_pos" << "  "
              << std::setw(9)  << "out_spt" << "  "
              << std::setw(9)  << "mot_vel" << "  "
              << std::setw(8)  << "torque"  << "\n"
              << std::string(62, '-') << "\n";

    // ---- Main loop ----
    while (g_running) {
        double now = now_s();
        double elapsed = now - t0;
        if (elapsed >= duration) break;

        // Heartbeat every 50ms
        if (now - last_hb >= 0.05) {
            motor.sendHeartbeat();
            last_hb = now;
        }

        auto frame = can.receive(2);  // 2ms timeout
        if (!frame) continue;
        if (frame->id != CAN_ID_FAST_FRAME || frame->data.size() < 8) continue;

        frames++;
        float pos;
        std::memcpy(&pos, frame->data.data(), 4);

        double dt = now - prev_time;
        if (dt < 0.0005) continue;
        float fdt = static_cast<float>(dt);

        // Velocity estimation (IIR filter)
        float vel_raw = (pos - prev_pos) / fdt;
        vel_filt = VEL_ALPHA * vel_raw + (1.0f - VEL_ALPHA) * vel_filt;
        prev_pos  = pos;
        prev_time = now;

        // ---- Trajectory generator (trapezoidal velocity profile) ----
        float to_go = target_rad - traj_pos;
        float dir   = (to_go > 0) ? 1.0f : -1.0f;
        float decel = traj_vel * traj_vel / (2.0f * MAX_ACCEL);

        if (std::abs(to_go) <= 0.001f && std::abs(traj_vel) < 0.1f) {
            traj_pos = target_rad;
            traj_vel = 0.0f;
        } else if (std::abs(to_go) <= decel) {
            traj_vel -= dir * MAX_ACCEL * fdt;
            if (dir * traj_vel < 0) traj_vel = 0.0f;
        } else {
            traj_vel += dir * MAX_ACCEL * fdt;
            traj_vel = std::clamp(traj_vel, -MAX_VEL, MAX_VEL);
        }
        traj_pos += traj_vel * fdt;

        // ---- PD + I controller tracking trajectory ----
        float pos_err = traj_pos - pos;
        float vel_err = traj_vel - vel_filt;

        integral += KI * pos_err * fdt;
        integral = std::clamp(integral, -INT_LIMIT, INT_LIMIT);

        float torque = KP * pos_err + KD * vel_err + integral;
        torque = std::clamp(torque, -TORQUE_LIMIT, TORQUE_LIMIT);

        motor.sendTorque(torque);

        // Print at ~4 Hz — display in output shaft degrees
        if (elapsed - last_print >= 0.25) {
            float out_pos = (pos - pos_start) / GEAR_RATIO * 180.0f / M_PI;
            float out_spt = (traj_pos - pos_start) / GEAR_RATIO * 180.0f / M_PI;
            float out_err = (pos_err) / GEAR_RATIO * 180.0f / M_PI;

            std::cout << std::fixed
                      << std::setw(4) << std::setprecision(1) << elapsed << "s  "
                      << std::showpos
                      << std::setw(7) << std::setprecision(2) << out_err << "deg  "
                      << std::noshowpos
                      << std::setw(7) << std::setprecision(1) << out_pos << "deg  "
                      << std::setw(7) << std::setprecision(1) << out_spt << "deg  "
                      << std::showpos
                      << std::setw(7) << std::setprecision(1) << vel_filt << "r/s  "
                      << std::setprecision(4) << torque << "Nm"
                      << std::noshowpos << "\n";
            last_print = elapsed;
        }
    }

    // ---- Stop ----
    std::cout << "\nStopping...\n";
    motor.sendTorque(0.0f);
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    motor.sendNMT(0x02);  // DAMPING

    double total = now_s() - t0;
    std::cout << "Frames: " << frames << " ("
              << static_cast<int>(frames / std::max(total, 0.1)) << " Hz)\n";

    auto fp = motor.sdoReadFloat(PARAM_POSITION);
    if (fp) {
        float out_final = (*fp - pos_start) / GEAR_RATIO * 180.0f / M_PI;
        float out_error = output_deg - out_final;
        std::cout << std::fixed << std::setprecision(1)
                  << "Output: " << out_final << " deg  (error: "
                  << std::showpos << out_error << " deg)\n"
                  << std::noshowpos;
    }

    std::cout << "Done.\n";
    return 0;
}
